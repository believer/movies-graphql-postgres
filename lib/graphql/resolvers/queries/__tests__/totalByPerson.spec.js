const totalByPerson = require('../totalByPerson')
const { verifyToken } = require('../../../../services/token')

jest.mock('../../../../services/token', () => ({
  verifyToken: jest.fn().mockImplementation(() =>
    Promise.resolve({
      id: 2,
    })
  ),
}))

describe('queries/totalByPerson', () => {
  let db

  beforeEach(() => {
    db = {
      query: jest.fn().mockReturnValue('dbData'),
    }
  })

  it('should verify token', () => {
    totalByPerson(null, { role: 'actor', ranked: 10 }, { db, token: 'test' })

    expect(verifyToken).toHaveBeenCalledWith('test')
  })

  it('should get how many movies actors are in sorted on ranking', async () => {
    try {
      await totalByPerson(null, { role: 'actor', ranked: 10 }, { db })

      expect(db.query.mock.calls[0][0]).toMatchSnapshot()
    } catch (e) {
      throw new Error(e)
    }
  })

  it('should get number of movies for a specific name', async () => {
    try {
      await totalByPerson(
        null,
        {
          role: 'actor',
          ranked: 10,
          name: 'Emma Watson',
        },
        { db }
      )

      expect(db.query.mock.calls[0][0]).toMatchSnapshot()
    } catch (e) {
      throw new Error(e)
    }
  })

  it('should return database data', async () => {
    const response = await totalByPerson(
      null,
      {
        role: 'actor',
        ranked: 10,
        name: 'Emma Watson',
      },
      { db }
    )

    expect(response).toEqual('dbData')
  })

  it('should handle errors', async () => {
    db.query.mockImplementation(() => Promise.reject('nope'))

    try {
      await totalByPerson(
        null,
        {
          role: 'actor',
          ranked: 10,
          name: 'Emma Watson',
        },
        { db }
      )
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
