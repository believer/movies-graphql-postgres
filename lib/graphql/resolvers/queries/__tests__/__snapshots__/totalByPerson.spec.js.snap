// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries/totalByPerson should get how many movies actors are in sorted on ranking 1`] = `
"
    SELECT * FROM (
      SELECT 
        people.name,
        people.number_of_movies,
        row_number() OVER (ORDER BY people.number_of_movies DESC) as ranking
      FROM 
        (
          SELECT 
            p.name,
            COUNT(p.name) as number_of_movies
          FROM 
            views AS v 
            INNER JOIN movies AS m ON m.id = v.movie_id 
            INNER JOIN cast_and_crew AS cac ON cac.movie_id = m.id 
            INNER JOIN people AS p ON cac.person_id = p.id 
          WHERE 
            v.user_id = 2
            AND cac.role_id = 2
          GROUP BY 1
        ) as people
    ) as ranked_people
    WHERE ranked_people.ranking <= $1
  "
`;

exports[`queries/totalByPerson should get number of movies for a specific name 1`] = `
"
    SELECT * FROM (
      SELECT 
        people.name,
        people.number_of_movies,
        row_number() OVER (ORDER BY people.number_of_movies DESC) as ranking
      FROM 
        (
          SELECT 
            p.name,
            COUNT(p.name) as number_of_movies
          FROM 
            views AS v 
            INNER JOIN movies AS m ON m.id = v.movie_id 
            INNER JOIN cast_and_crew AS cac ON cac.movie_id = m.id 
            INNER JOIN people AS p ON cac.person_id = p.id 
          WHERE 
            v.user_id = 2
            AND cac.role_id = 2
          GROUP BY 1
        ) as people
    ) as ranked_people
    WHERE ranked_people.name = $1
  "
`;
